redwood/src/lib.rs

```diff
@@ -217,6 +255,51 @@ mod tests {
         assert_eq!(format!("{}", cert.fingerprint()), fingerprint);
     }
 
+    #[test]
+    fn test_source_cert_has_accepted_key() {
+        // Certificate with valid key
+        let (public_key, _secret_key, _fingerprint) =
+            generate_source_key_pair(PASSPHRASE, "foo@example.org").unwrap();
+        let good_cert = Cert::from_str(&public_key).unwrap();
+
+        // Certificate with unsupported key
+        let bad_cert = Cert::from_bytes(BAD_KEY.as_bytes()).unwrap();
+        assert_eq!(BAD_KEY_FINGERPRINT, bad_cert.fingerprint().to_string());
+
+        // Expected num supported keys for each type of certificate
+        let certs_and_results = vec![(good_cert, 1), (bad_cert, 0)];
+
+        for (cert_collection, expected_supported_keys) in certs_and_results {
+            let mut actual_supported_keys = 0;
+            for _key in cert_collection
+                .keys()
+                .with_policy(&StandardPolicy::new(), None)
+                .supported()
+                .alive()
+                .revoked(false)
+                .for_storage_encryption()
+            {
+                actual_supported_keys += 1;
+            }
+            assert_eq!(expected_supported_keys, actual_supported_keys);
+        }
```

Comment:
(It may be fine to just skip to testing `encrypt_message` as a whole, in which case this test function can be shorter n sweeter)

Reply:
I would actually do the opposite, *just* test `encrypt_message` correctly fails instead of trying to count the number of supported keys, which requires duplicating the supported keys logic. If we want to test the underlying iterator then I think we should be splitting it out to a separate function.